test/junit/scala/tools/nsc/backend/jvm/NestedClassesCollectorTest.scala

```diff
@@ -0,0 +1,115 @@
+    ref("Lp/Kl.Ne.In;", List("p/Kl", "p/Kl$Ne", "p/Kl$Ne$In"))
+    ref("LA<*>;", List("A"))
+    ref("LA<**+[I[JTFoo;-TBar;LB;*>;", List("A", "B"))
+    ref("Lp/A<[I[LTBoo<*>;-[JTFoo;-TBar;Lp/B<[J+[Lp/C;>.N<+TT;*Lp/D;>;*>;", List("TBoo", "p/A", "p/B", "p/B$N", "p/C", "p/D"))
+    ref("Lp/A<[I[Lp/B<*>;>;", List("p/A", "p/B"))
+    ref("Lp/A<Lp/B;>.C<Lp/D;>.E;", List("p/A", "p/A$C", "p/A$C$E", "p/B", "p/D"))
+
+    ref("[I", Nil)
+    ref("[[[LA;", List("A"))
+    ref("[[[LA<**+[I-[LB;>;", List("A", "B"))
+  }
+
+  @Test
+  def classSignatures(): Unit = {
+    def cls(sig: String, expect: List[String]) = {
+      c.visitClassSignature(sig)
+      assertEquals(inners, expect)
+    }
+
+    cls("LA;", List("A"))
+    cls("LA;LB;", List("A", "B"))
+    cls("Lp/a/A;Lp/B;", List("p/B", "p/a/A"))
+    cls("<T:>LA;", List("A"))
+    cls("<T:LA;:[I:TU;:[TV;>LB;", List("A", "B"))
+    cls("<T:LA;:[Lp/B<+Lp/C;>;:TU;:[TV;>LC;", List("A", "C", "p/B", "p/C"))
+    cls("<T::TT;>LA;", List("A"))
+    cls("<T:LA;>LB;", List("A", "B")) // one type parameter T with class bound A
+    cls("<P:TT;>LA;", List("A")) // one type parameter
+
+    // Missing ClassBound without an interface bound. Probably the grammar only allows those by
+    // accident. Our parser doesn't. https://stackoverflow.com/q/44284928
+    // cls("<T:U:>LA;", List("A"))
+    // cls("<T:L:>LA;", List("A"))
+    // cls("<T:LA:>LB;", List("B")) // two type parameters, T and LA
+    // cls("<P:TT:>LA;", List("A")) // two type parameters
+    // cls("<T:U:LA;>LB;", List("A", "B"))
+  }
+
+  @Test
+  def methodSignatures(): Unit = {
+    def met(sig: String, expect: List[String]) = {
+      c.visitMethodSignature(sig)
+      assertEquals(inners, expect)
+    }
+
+    // type parameters implementation is the same as for class signatures, so only basic testing here
+    met("()V", Nil)
+    met("(BJI)Z", Nil)
+    met("(IJLp/A;Z)Lp/B;", List("p/A", "p/B"))
+    met("<X:LA;:[I:TU;:[TV;Y:[I:LB<+LC;>;>([I[[[LD<**>;)TT;", List("A", "B", "C", "D"))
+    met("(LA;ITT;)I^LB;", List("A", "B"))
+    met("()I^TT;^Lp/A<**+[[Lp/B;>;^TBA;", List("p/A", "p/B"))
+    met("()V^TT;", Nil)
+  }
+
+  @Test
+  @Ignore("manually run test")
+  def rtJar(): Unit = {
+    import java.nio.file._
+    import scala.collection.JavaConverters._
+    val zipfile = Paths.get("/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/rt.jar")
+    val fs = FileSystems.newFileSystem(zipfile, null)
+    val root = fs.getRootDirectories.iterator().next()
+    val contents = Files.walk(root).iterator().asScala.toList
+    for (f <- contents if Files.isRegularFile(f) && f.getFileName.toString.endsWith(".class")) {
+      val classNode = AsmUtils.classFromBytes(Files.readAllBytes(f))
+      c.visitClassSignature(classNode.signature)
+      classNode.methods.iterator().asScala.map(_.signature).foreach(c.visitMethodSignature)
+      classNode.fields.iterator().asScala.map(_.signature).foreach(c.visitFieldSignature)
+    }
+  }
```

Comment:
> Needs investigation.

I'm almost certain that the errant classfile was built with a buggy compiler I created when trying to optimize the creation of generic signatures.